main/src/library/Reducible.flix

```diff
@@ -0,0 +1,206 @@
+/*
+ *  Copyright 2022 Jakob Schneider Villumsen
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+// This type class is inspired by the Cats implementation of Reducible:
+// https://github.com/typelevel/cats/blob/main/core/src/main/scala/cats/Reducible.scala
+
+///
+/// A type class for types that can be reduced to a summary value.
+///
+/// `Reducible` is like a non-empty `Foldable` and may only be implemented on non-empty data structures.
+///
+///
+pub lawless class Reducible[t: Type -> Type] {
+
+    ///
+    /// Left-associative reduction of a structure.
+    /// Applies `g` to the initial element of `t` and combines it
+    /// with the remainder of `t` using `f` going from left to right.
+    ///
+    pub def reduceLeftTo(f: (b, a) -> b & ef1, g: a -> b & ef2, t: t[a]): b & (ef1 and ef2)
+
+    ///
+    /// Right-associative reduction of a structure.
+    /// Applies `g` to the initial element of `t` and combines it
+    /// with the remainder of `t` using `f` going from right to left.
+    ///
+    pub def reduceRightTo(f: (a, b) -> b & ef1, g: a -> b & ef2, t: t[a]): b & (ef1 and ef2)
+
+    ///
+    /// Left-associative reduction on `t` using `f`.
+    ///
+    pub def reduceLeft(f: (a, a) -> a & ef, t: t[a]): a = ???
+
+    ///
+    /// Right-associative reduction on `t` using `f`.
+    ///
+    pub def reduceRight(f: (a, a) -> a & ef, t: t[a]): a = ???
+
+    ///
+    /// Reduce `t` using the derived `SemiGroup` instance.
+    ///
+    pub def reduce(t: t[a]): a with SemiGroup[a] = ???
+
+    ///
+    /// Applies `f` to each element of `t` and combines them using the derived `SemiGroup` instance.
+    ///
+    pub def reduceMap(f: a -> b & ef, t: t[a]): b with SemiGroup[b] = ???
+
+    ///
+    /// Left-associative fold of a structure.
+    /// Applies `f` to a start value `s` and all elements in `t` going from left to right.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, s: b, t: t[a]): b = ???
+
+    ///
+    /// Right-associative fold of a structure.
+    /// Applies `f` to a start value `s` and all elements in `t` going from right to left.
+    ///
+    pub def foldRight(f: (a, b) -> b & ef, s: b, t: t[a]): b = ???
+
+    ///
+    /// Returns the first element of `t`.
+    ///
+    pub def head(t: t[a]): a = ???
+
+    ///
+    /// Returns the tail of `t`.
+    ///
+    pub def tail(t: t[a]): t[a] = ???
+
```

Comment:
+ init

Reply:
What type of `init`? I found this in `Chain`

```flix
    ///
    /// Returns the subchain of `c` without the last element.
    /// Returns `None` if the chain `c` is empty.
    ///
    pub def init(c: Chain[a]): Option[Chain[a]]
```
but also this in `Array`
```flix

    ///
    /// Build an array of length `len` by applying `f` to the successive indices.
    ///
    @Time(time(f) * len) @Space(space(f) * len)
    pub def init(f: Int32 -> a & ef, len: Int32): Array[a] & Impure
```

Reply:
I think it has to be list.

Reply:
Sorry I was being unclear. I meant to ask about what arguments it should accept, a function `f` and a length / integer?

Reply:
No argument. `init` is everything but the first element.

Reply:
its the dual of tail

Reply:
I would keep `head` and `last` together and `init` and `tail` together.